File: mobile/android/app/src/main/kotlin/com/slopssaloon/omen/app/feature/api/MinVersionGateClient.kt
package com.slopssaloon.omen.app.feature.api

import org.json.JSONObject

/**
 * O7 — server-driven minimum-version gate. iOS mirror:
 * `App/Api/MinVersionGateClient.swift`.
 *
 * Deliberately unauthenticated: the check runs at launch, before sign-in, so it cannot
 * depend on a bearer token.
 */
sealed interface MinVersionGateResult {
    data object Ok : MinVersionGateResult
    data class UpdateRequired(val minimumVersion: String) : MinVersionGateResult

    /**
     * Any failure — network, non-2xx, unreadable body. Callers MUST treat this identically
     * to [Ok]. This gate exists to block a known-bad build, never to add a new way to get
     * locked out of the app.
     */
    data object Unavailable : MinVersionGateResult
}

fun interface MinVersionGateChecking {
    suspend fun check(platform: String, currentVersion: String): MinVersionGateResult
}

class MinVersionGateClient(
    apiBaseUrl: String,
    private val fetcher: OmenHttpFetcher = OkHttpFetcher(),
) : MinVersionGateChecking {
    private val base = apiBaseUrl.trimEnd('/')

    override suspend fun check(platform: String, currentVersion: String): MinVersionGateResult {
        val url = "$base/api/system/min-version?platform=$platform&version=$currentVersion"
        // Empty access token: the route is public and ignores Authorization. Reusing the
        // shared fetcher seam keeps the test fakes identical to every other API test.
        //
        // `runCatching` rather than trusting the fetcher's own handling: `OkHttpFetcher`
        // catches `IOException` only, and builds the `Request` outside that try — a
        // malformed base URL throws `IllegalArgumentException` instead. This check runs
        // unconditionally at launch, before sign-in, so an escaping exception would crash
        // the app rather than fail open, which is the one thing this gate must never do.
        // The iOS twin already collapses every error; this makes the guarantee symmetric.
        val (status, payload) = runCatching { fetcher.fetch(url, "GET", "", null) }.getOrNull()
            ?: return MinVersionGateResult.Unavailable

        if (status !in 200..299) return MinVersionGateResult.Unavailable

        val json = runCatching { JSONObject(payload) }.getOrNull()
            ?: return MinVersionGateResult.Unavailable

        if (!json.optBoolean("update_required", false)) return MinVersionGateResult.Ok

        val minimumVersion = json.optString("minimum_version").takeIf { it.isNotBlank() && it != "null" }
            ?: return MinVersionGateResult.Unavailable

        return MinVersionGateResult.UpdateRequired(minimumVersion)
    }
}

File: mobile/ios/OmenIOS/OmenIOS/App/Api/MinVersionGateClient.swift
import Foundation

/// O7 — server-driven minimum-version gate.
///
/// Mirrors `OmenApiClient`'s transport shape but is deliberately unauthenticated: the check
/// must run before sign-in, at launch, so it cannot depend on a bearer token.
struct MinVersionStatus: Decodable, Equatable {
    let status: String
    let updateRequired: Bool
    let minimumVersion: String?

    enum CodingKeys: String, CodingKey {
        case status
        case updateRequired = "update_required"
        case minimumVersion = "minimum_version"
    }
}

/// Collapsed to what the gate view actually needs to render. Any failure — network, decode,
/// non-2xx — resolves to `.unavailable`, which the caller must treat identically to `.ok`.
/// This gate exists to block a known-bad build, not to add a new way to get locked out.
enum MinVersionGateResult: Equatable {
    case ok
    case updateRequired(minimumVersion: String)
    case unavailable
}

protocol MinVersionGateChecking {
    func check(platform: String, currentVersion: String) async -> MinVersionGateResult
}

struct MinVersionGateClient: MinVersionGateChecking {
    private let baseURL: URL
    private let fetcher: OmenHTTPFetching
    private let timeout: TimeInterval

    init(baseURL: URL, fetcher: OmenHTTPFetching = URLSession.shared, timeout: TimeInterval = 10) {
        self.baseURL = baseURL
        self.fetcher = fetcher
        self.timeout = timeout
    }

    func check(platform: String, currentVersion: String) async -> MinVersionGateResult {
        var components = URLComponents(
            url: baseURL.appendingPathComponent("api/system/min-version"),
            resolvingAgainstBaseURL: false
        )
        components?.queryItems = [
            URLQueryItem(name: "platform", value: platform),
            URLQueryItem(name: "version", value: currentVersion),
        ]
        guard let url = components?.url else { return .unavailable }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = timeout
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await fetcher.data(for: request)
        } catch {
            return .unavailable
        }

        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            return .unavailable
        }
        guard let decoded = try? JSONDecoder().decode(MinVersionStatus.self, from: data) else {
            return .unavailable
        }
        guard decoded.updateRequired, let minimumVersion = decoded.minimumVersion else {
            return .ok
        }
        return .updateRequired(minimumVersion: minimumVersion)
    }
}

/// Reads the running build's own version, the same value sent to the gate endpoint.
enum AppVersionProvider {
    static var current: String {
        (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "0.0.0"
    }
}
